src/Features/Core/Portable/EditAndContinue/EditAndContinueWorkspaceService.cs

```diff
@@ -387,18 +384,38 @@ public Task<bool> HasChangesAsync(Solution solution, SolutionActiveStatementSpan
                 editSession.StorePendingUpdate(solution, solutionUpdate);
             }
 
-            // clear emit/apply diagnostics reported previously:
-            _emitDiagnosticsUpdateSource.ClearDiagnostics();
+            // Note that we may return empty deltas if all updates have been deferred.
+            // The debugger will still call commit or discard on the update batch.
+            return (solutionUpdate.Summary, solutionUpdate.Deltas, ToDiagnosticData(solution, solutionUpdate.Diagnostics));
+        }
 
-            // report emit/apply diagnostics:
-            foreach (var (projectId, diagnostics) in solutionUpdate.Diagnostics)
+        private static ImmutableArray<DiagnosticData> ToDiagnosticData(Solution solution, ImmutableArray<(ProjectId ProjectId, ImmutableArray<Diagnostic> Diagnostics)> diagnosticsByProject)
+        {
+            var _ = ArrayBuilder<DiagnosticData>.GetInstance(out var result);
+
+            foreach (var (projectId, diagnostics) in diagnosticsByProject)
             {
-                _emitDiagnosticsUpdateSource.ReportDiagnostics(_workspace, solution, projectId, diagnostics);
+                var project = solution.GetProject(projectId);
+
+                foreach (var diagnostic in diagnostics)
+                {
+                    var document = solution.GetDocument(diagnostic.Location.SourceTree);
+                    if (document != null)
+                    {
+                        result.Add(DiagnosticData.Create(diagnostic, document));
+                    }
+                    else if (project != null)
+                    {
+                        result.Add(DiagnosticData.Create(diagnostic, project));
+                    }
+                    else
+                    {
+                        result.Add(DiagnosticData.Create(diagnostic, solution.Options));
```

Comment:
how can this be hit?  if we have ` foreach (var (projectId, diagnostics) in diagnosticsByProject)`, and diagonsticsByProject always has a non-null proejctId, we should *always* find a project, right?  

basically, i can't tell if this is code that is just trying to be resilient to crazyness, or if these are expected cases and the types aren't properly annotated.  in either event, docs would be very helpful.

Reply:
Good catch. Turns out we actually don't have solution diagnostics. We always associate the error with a project. This can be simplified.